packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js

```diff
@@ -442,11 +444,67 @@ export default {
           }
         }
 
+        gatherJSXDependencies(currentScope);
+
         for (const childScope of currentScope.childScopes) {
           gatherDependenciesRecursively(childScope);
         }
       }
 
+      gatherJSXDependencies(node);
+
+      function gatherJSXDependencies(rootNode) {
+        const JSXComponents = [];
+
+        estraverse.traverse(rootNode, {
+          enter: function(currentNode) {
+            const {name, type} = currentNode;
+            if (isJSXComponent(name, type) && !JSXComponents.includes(name)) {
+              JSXComponents.push(name);
+            }
+          },
+          // a list of all JSX keys and its attributes that can contain
+          // other JSX elements and identifiers
+          keys: {
+            JSXElement: [
+              'type',
+              'openingElement',
+              'closingElement',
+              'children',
+            ],
+            JSXFragment: ['children'],
+            JSXAttribute: ['name', 'value'],
+            JSXIdentifier: ['name', 'type'],
+            JSXOpeningElement: ['name', 'attributes'],
+            JSXClosingElement: ['name'],
+            JSXNamespacedName: ['namespace', 'name'],
+            JSXSpreadAttribute: ['argument'],
+            JSXMemberExpression: ['object'],
+            JSXExpressionContainer: ['expression'],
+          },
+          fallback: () => [], // just skip all unknown nodes, otherwise there will be an exception
+        });
+
+        function isHTMLElement(name) {
+          return /^[a-z]/.test(name);
+        }
+
+        function isJSXComponent(name, type) {
+          return (
+            type === 'JSXIdentifier' &&
+            name !== 'Fragment' &&
```

Comment:
It doesn't seem right that we need a special case here. I think we want to rely on the same logic that we use for identifiers in general. If they're coming from a "pure scope" — e.g. from within the component function — then we want them to be deps. But if they're coming from outside — e.g. from the top-level import — then they shouldn't be dependencies. We already have this logic for regular identifiers, so we need to be able to reuse that somehow.

Reply:
So, I have implemented a solution to this problem, and the common logic of the solution is as follows:

- for the reason that our dependencies must came from pure scopes, I went through the set of pure scopes and picked up all the variables that are declared in them. This list will include function arguments, variables declarations, functions declarations etc.
- I have added the `isPureScopeDependency` predicate, which checks if a JSX component is in the pure scope variable list. If this is true, then the component is included in the dependency list, if not, then it does not.
- I've removed the special conditions that checked if the JSXIdentifier is equal to `React`, `Fragment`, etc. Now these JSXIdentifiers will not be included in the list of dependencies automatically, as well as all other JSXIdentifiers that are not among the pure scopes variable list.

I couldn't directly reuse the logic with regular identifiers, because it is based on references and their resolved variables. JSXIdentifiers have no variable references in its interface and I didn't find a way to add it to them.
Therefore, I tried to implement conceptually similar logic, but based on variables from pure scopes.